contrib/qpl-cmake/CMakeLists.txt

```diff
@@ -0,0 +1,362 @@
+                           PRIVATE ${QPL_SRC_DIR}/isal/include
+                           PUBLIC ${QPL_SRC_DIR}/isal/igzip)
+
+set_target_properties(isal PROPERTIES
+                      CXX_STANDARD 11
+                      C_STANDARD 99)
+
+target_compile_options(isal PRIVATE
+                       "$<$<C_COMPILER_ID:GNU>:${QPL_LINUX_TOOLCHAIN_REQUIRED_FLAGS}>"
+                       "$<$<C_COMPILER_ID:MSVC>:/WX;${DISABLED_WARNINGS_FLAGS};${QPL_WINDOWS_TOOLCHAIN_REQUIRED_FLAGS}>"
+                       "$<$<CONFIG:Debug>:>"
+                       "$<$<CONFIG:Release>:>")
+
+target_compile_options(isal_asm PUBLIC "-I${QPL_SRC_DIR}/isal/include/"
+                       PUBLIC "-I${QPL_SRC_DIR}/isal/igzip/"
+                       PUBLIC "-I${QPL_SRC_DIR}/isal/crc/"
+                       PUBLIC "-DHAVE_AS_KNOWS_AVX512"
+                       PUBLIC "-DAS_FEATURE_LEVEL=10"
+                       PUBLIC "-DQPL_LIB")
+
+if (SANITIZE STREQUAL "undefined")
+    get_target_property(target_options isal_asm COMPILE_OPTIONS)
+    list(REMOVE_ITEM target_options "-fno-sanitize=undefined")
+    set_property(TARGET isal_asm PROPERTY COMPILE_OPTIONS ${target_options})
+endif()
+
+target_compile_definitions(isal PUBLIC
+                           QPL_LIB
+                           NDEBUG)
+
+# [SUBDIR]core-sw
+enable_language(C)
+
+#
+# Create avx512 library
+#
+#set(CMAKE_INCLUDE_CURRENT_DIR ON)
+
+# Find Core Sources
+file(GLOB SOURCES
+     ${QPL_SRC_DIR}/core-sw/src/checksums/*.c
+     ${QPL_SRC_DIR}/core-sw/src/filtering/*.c
+     ${QPL_SRC_DIR}/core-sw/src/other/*.c
+     ${QPL_SRC_DIR}/core-sw/src/compression/*.c)
+
+file(GLOB DATA_SOURCES
+     ${QPL_SRC_DIR}/core-sw/src/data/*.c)
+
+# Create library
+add_library(qplcore_avx512 OBJECT ${SOURCES})
+
+target_compile_definitions(qplcore_avx512 PRIVATE PLATFORM=2)
+
+target_include_directories(qplcore_avx512
+                           PUBLIC $<BUILD_INTERFACE:${QPL_SRC_DIR}/core-sw/include>
+                           PUBLIC $<BUILD_INTERFACE:${QPL_SRC_DIR}/core-sw/src/include>
+                           PUBLIC $<BUILD_INTERFACE:${QPL_SRC_DIR}/core-sw/src/compression/include>
+                           PRIVATE $<TARGET_PROPERTY:isal,INTERFACE_INCLUDE_DIRECTORIES>)
+
+set_target_properties(qplcore_avx512 PROPERTIES
+        $<$<C_COMPILER_ID:MSVC>:C_STANDARD 18>
+        $<$<C_COMPILER_ID:GNU>:C_STANDARD 17>)
+
+target_link_libraries(qplcore_avx512 ${CMAKE_DL_LIBS} isal)
+
+if (WIN32)
+    target_compile_options(qplcore_avx512
+                           PRIVATE ${QPL_WINDOWS_TOOLCHAIN_REQUIRED_FLAGS}
+                           PRIVATE /arch:AVX512
+                           PRIVATE "$<$<CONFIG:Debug>:>"
+                           PRIVATE "$<$<CONFIG:Release>:-O2>")
+else ()
+    target_compile_options(qplcore_avx512
+                           PRIVATE ${QPL_LINUX_TOOLCHAIN_REQUIRED_FLAGS}
+                           PRIVATE -march=skylake-avx512
```

Comment:
Two libraries:qplcore_avx512/qplcore_px for SW fallback will be created which are implemented by AVX512 and non-AVX512 instructions respectively.
The upper level QPL API will check SIMD capabilities of the target system at runtime and decide to call AVX512 function or non-AVX512 function.
Hence, here we don't need put qplcore_avx512 under an ENABLE_AVX512 CMake switch. Actually, if we do that, some undefined symbols errors would happen because both of AVX512 function and non-AVX512 function are referenced by QPL API.